docs_src/design/adr/0006-Service-Registry.md

```diff
@@ -0,0 +1,183 @@
+started.
+
+This release actually preceded usage of TOML for configuration files, and instead just used a flat key/value format,
+with keys converted from legacy Java property names (e.g. meta.db.device.url ) to Camel[Pascal]/Case (e.g. MetaDeviceServiceURL).
+
+I chose the config key mentioned above on purpose:
+
+```
+MetaDeviceURL = "http://edgex-core-metadata:48081/api/v1/device"
+```
+
+Not only did this config key provide the address of core metadata, it also provided the path of a specific REST endpoint. In later releases
+of EdgeX, the address of the service and the specific endpoint paths were de-coupled. Instead of following the Service Name design (which
+was finalized two months earlier), the initial implementation followed the legacy Java implementation and initialized its service clients
+for each required REST endpoint (belonging to another EdgeX service) directly from the associated *URL config key read from Consul (if
+enabled) or directly from the configuration file.
+
+The shared client initialization code also created an Endpoint monitor goroutine and passed it a go channel channel used by the service
+to receive updates to the REST API endpoint URL. This monitor goroutine effectively polled Consul every 15s (this became configurable in
+later versions) for the client's service address and if a change was detected, would write the updated endpoint URL to the given channel,
+effectively ensuring that the service started using the new URL.
+
+It wasn't till late in the Geneva development cycle that I noticed log messages which made me aware of the fact that every one of our
+services was making a REST call to check the address of a service endpoint every 15s, for **every** REST endpoint it used! An issue was
+filed (https://github.com/edgexfoundry/edgex-go/issues/2594), and the client monitoring was removed as part of the Geneva 1.2.1 release.
+
+## Problem Statement
+The fundamental problem with the existing implementations (as decribed above), is that there's too much duplication of configuration across
+services. For instance, Core Data's service port can easily be changed by passing the environment variable SERVICE_PORT to the service on
+startup. This overrides the configuration read from the configuration provider, and will cause Core Data to listen on the new port,
+however it has no impact on any services which use Core Data, as the client config for each is read from the configuration provider (excluding
+security-proxy-setup).
+
+This means in order to change a service port, environment variable overrides need to set for every client service as well as security-proxy-setup
+(if required).
+
+## Decision
+Update the core, support, and security-proxy-setup to use go-mod-registry's ```Client.GetServiceEndpoint``` method on startup to determine (a) if
+a service dependency is available and (b) use the returned address information to initialize client endpoints (or setup the correct route in the
+case of proxy-setup).
```

Comment:
What about additional metadata such as:
- The URL path described earlier
- Which protocol to use to connect (tcp, http, https, etc...)

Where did this metadata go.  It used to be part of the original spec:
```
MetaDeviceURL = "http://edgex-core-metadata:48081/api/v1/device"
```

Reply:
The URL is constructed using a set of constants which define the available endpoints. In the original Java code (and the first version of the re-written Go services) the actual endpoint paths were defined in the configuration files which made no sense at all.  Just because an endpoint path was added to a configuration file, didn't mean the service being called actually implements the given endpoint.

Also with regards to protocol, all of our endpoints only support HTTP. For instances where additional protocols are supported, additional configuration is required.

Reply:
Marking this resolved. If there's still an issue, feel free to re-open with relevant details.